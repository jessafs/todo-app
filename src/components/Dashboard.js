import React, { useState, useEffect } from "react";
import Modal from "./modal"
import Accordion from "./Accordion";
import { v4 as uuid } from 'uuid';
import { FaTrashAlt, FaPenAlt, FaPlus } from "react-icons/fa";
import "./dashboard.css";
function Dashboard() {
  const initialState = {
    todo: {
      id: '',
      title: '',
      content: ''
    }
  }
  const [modalShow, setModalShow] = useState(false);
  const [btnChange, setBtnChange] = useState(false);
  const [data, setData] = useState(initialState);
  const [todoList, setTodoList] = useState([]);
  const handleShow = () => {
    setModalShow(true)
    setBtnChange(true)
  }
  const handleClose = () => setModalShow(false)
  const isEmpty = Object.values(data.todo).some(x => x === null || x === '');


  const handleInputTodo = (e) => {
    const { name, value } = e.target
    setData({ todo: { ...data.todo, [name]: value } })//id: uuid() 
  }
  const handleGetTodoList = () => {
    if (isEmpty && todoList.length === 0) {
      setTodoList([])
    } else {
      setTodoList([...todoList, data.todo])
    }
  }

  const handleAddTodo = () => {
    const { title, content } = data.todo ?? {}
    const id = uuid()
    if (!title || !content) {
      alert("All fields are required")
    } else {
      const newData = {
        id,
        title,
        content
      }
      setTodoList([...todoList, newData])
      handleClose()
      setData({
        ...data,
        todo: {
          id: "",
          title: "",
          content: ""
        }
      })
    }
  }

  const handleCancel = () => {
    handleClose()
    setData({
      ...data,
      todo: {
        id: "",
        title: "",
        content: ""
      }
    })
  }
  const handleRemoveList = (id) => {
    if (!id) {
      alert("no id")
    } else {
      let txt = ''
      if (window.confirm("Are you sure you want to delete this todo?")) {
        txt = "Successfully deleted!";
        const newData = todoList.filter((card) => card.id !== id)
        setTodoList(newData)
      } else {
        txt = "You pressed Cancel!";
      }

    }

  }
  const handleOpenUpdateTodo = (id) => {
    if (!id) {
      alert("Not found")
    } else {
      const dataToUpdate = todoList.filter((data) => {
        if (data.id === id) {
          return data
        }
      })
      setBtnChange(false)
      setData({ todo: { ...dataToUpdate[0] } })
      setModalShow(true)
    }
  }
  const handleUpdatetodo = () => {
    const { title, content, id } = data.todo ?? {}
    if (!title || !content) {
      alert("All fields are required")
    } else {
      const newData = todoList.map((item) => {
        if (item.id === id) {
          return data.todo
        } else {
          return item
        }
      })
      setTodoList(newData)
      handleClose()
      setData({
        ...data,
        todo: {
          id: "",
          title: "",
          content: ""
        }
      })
    }
  }

  useEffect(() => {
    handleGetTodoList()
  }, []);
  return (
    <div className="container">
      <h1 data-testid="header">TODO LIST</h1>
      <div className="btnContainer">
        <button className="btnAdd" onClick={handleShow}>  <FaPlus /> Add</button>
      </div>
      <Modal
        show={modalShow}
        btnChange={btnChange}
        handleClose={handleClose}
        handleInputTodo={handleInputTodo}
        handleAddTodo={handleAddTodo}
        handleCancel={handleCancel}
        handleUpdatetodo={handleUpdatetodo}
        data={data}
      />
      {/* <div class="contentList">
        {todoList.length === 0 ? <h1 class="noData">No Available Data!!!</h1> :
          <div class="parent">
            {
              todoList.map((todo, i) => {
                const { title, content, id } = todo
                const formattedTitle = title.charAt(0).toUpperCase() + title.slice(1)
                return (
                  < div class="row" key={i} >
                    <div class="column">
                      <small className="cardTitle">{formattedTitle}</small>
                      <div class="card-content" >
                        <p>{content}</p>
                        <div class="card-footer">
                          <FaTrashAlt class="deleteIcon" onClick={() => handleRemoveList(id)} />&nbsp;&nbsp;
                          <FaPenAlt class="editIcon" onClick={() => handleOpenUpdateTodo(id)} />
                        </div>
                      </div>
                    </div>
                  </div>
                )
              })
            }
          </div>
        }
      </div> */}
      {/* accordion */}
      <div className="contentList">
        {todoList.length === 0 ? <h1 class="noData">No Available Data!!!</h1> :
          <div>
            {todoList.map(({ title, content, id }) => (
              <Accordion
                id={id}
                title={title}
                content={content}
                handleRemoveList={handleRemoveList}
                handleOpenUpdateTodo={handleOpenUpdateTodo}
              />
            ))}
          </div>
        }
      </div>

    </div >
  );
}

export default Dashboard;
