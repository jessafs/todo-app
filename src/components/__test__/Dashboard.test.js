import React from "react";
import Dashboard from "../Dashboard";
import renderer from "react-test-renderer";
// import { render } from "@testing-library/react";

import "@testing-library/jest-dom/extend-expect";

test("renderes user input", () => {
  // create a snapshot
  const tree = renderer.create(
    <Dashboard
      users={[]}
      handleShow={() => {}}
      isListVisible={true}
      currentIndex={0}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot()
});
test("renders todo input", () => {
    // create a snapshot
    // tree is a representation of a html output of the compoenent 
    const tree = renderer.create(
      <Dashboard
        users={[{title:"First todo",content:"The quuick brown fix jump over the lazy dog",id:'1'}]}
        handleShow={() => {}}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot()
  });

// App.test.js
// import { render, screen } from '@testing-library/react';
// import App from './App';


