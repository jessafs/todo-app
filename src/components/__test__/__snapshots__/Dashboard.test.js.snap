// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderes user input 1`] = `
<div
  className="container"
>
  <h1
    data-testid="header"
  >
    TODO LIST
  </h1>
  <div
    className="btnContainer"
  >
    <button
      className="btnAdd"
      onClick={[Function]}
    >
        
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        />
      </svg>
       Add
    </button>
  </div>
  <div
    className="modal display-none"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          className="close"
          onClick={[Function]}
        >
          ×
        </button>
        <h4>
          Edit Todo
        </h4>
      </div>
      <div
        className="modal-body"
      >
        <label>
          Title
        </label>
        <input
          className="topicTitle"
          name="title"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label>
          Content
        </label>
        <textarea
          className="topicContent"
          name="content"
          onChange={[Function]}
          value=""
        />
      </div>
      <div
        className="modal-footer"
      >
        <div />
        <button
          className="cancelbtn"
          onClick={[Function]}
        >
          Cancel
        </button>
        <button
          className="savebtn"
          onClick={[Function]}
        >
          Update
        </button>
      </div>
    </div>
  </div>
  <div
    className="contentList"
  >
    <h1
      class="noData"
    >
      No Available Data!!!
    </h1>
  </div>
</div>
`;

exports[`renders todo input 1`] = `
<div
  className="container"
>
  <h1
    data-testid="header"
  >
    TODO LIST
  </h1>
  <div
    className="btnContainer"
  >
    <button
      className="btnAdd"
      onClick={[Function]}
    >
        
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        />
      </svg>
       Add
    </button>
  </div>
  <div
    className="modal display-none"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          className="close"
          onClick={[Function]}
        >
          ×
        </button>
        <h4>
          Edit Todo
        </h4>
      </div>
      <div
        className="modal-body"
      >
        <label>
          Title
        </label>
        <input
          className="topicTitle"
          name="title"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label>
          Content
        </label>
        <textarea
          className="topicContent"
          name="content"
          onChange={[Function]}
          value=""
        />
      </div>
      <div
        className="modal-footer"
      >
        <div />
        <button
          className="cancelbtn"
          onClick={[Function]}
        >
          Cancel
        </button>
        <button
          className="savebtn"
          onClick={[Function]}
        >
          Update
        </button>
      </div>
    </div>
  </div>
  <div
    className="contentList"
  >
    <h1
      class="noData"
    >
      No Available Data!!!
    </h1>
  </div>
</div>
`;
